fsw/cfe-core/src/inc/cfe_msg.h

```diff
@@ -0,0 +1,220 @@
+**
+**  Unless required by applicable law or agreed to in writing, software
+**  distributed under the License is distributed on an "AS IS" BASIS,
+**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**  See the License for the specific language governing permissions and
+**  limitations under the License.
+*/
+
+/******************************************************************************
+ * Message access APIs
+ */
+
+#ifndef _cfe_msg_
+#define _cfe_msg_
+
+/*
+ * Includes
+ */
+#include "common_types.h"
+#include "cfe_mission_cfg.h"
+#include "cfe_sb.h"
+
+/** @defgroup CFEAPIMSGChecksum cFE Message Checksum Control APIs
+ * @{
+ */
+
+/*****************************************************************************/
+/**
+** \brief Gets the checksum field from a software bus message.
+**
+** \par Description
+**          This routine gets the checksum (or other message integrity check
+**          value) from a software bus message.  The contents and location of
+**          this field will depend on the underlying implementation of software
+**          bus messages.  It may be a checksum, a CRC, or some other algorithm.
+**          Users should not call this function as part of a message integrity
+**          check (call #CFE_SB_ValidateChecksum instead).
+**
+** \par Assumptions, External Events, and Notes:
+**          - If the underlying implementation of software bus messages does not
+**            include a checksum field, then this routine will return a zero.
+**
+** \param[in]  MsgPtr      A pointer to the buffer that contains the software bus message.
+**                         This must point to the first byte of the message header.
+**
+** \return The checksum included in the software bus message header (if present), otherwise,
+**         returns a checksum value of zero.
+**/
+uint16 CFE_MSG_GetChecksum(CFE_SB_MsgPtr_t MsgPtr);
+
+/*****************************************************************************/
+/**
+** \brief Calculates and sets the checksum of a software bus message
+**
+** \par Description
+**          This routine calculates the checksum of a software bus message according
+**          to an implementation-defined algorithm.  Then, it sets the checksum field
+**          in the message with the calculated value.  The contents and location of
+**          this field will depend on the underlying implementation of software bus
+**          messages.  It may be a checksum, a CRC, or some other algorithm.
+**
+** \par Assumptions, External Events, and Notes:
+**          - If the underlying implementation of software bus messages does not
+**            include a checksum field, then this routine will do nothing.
+**
+** \param[in]  MsgPtr      A pointer to the buffer that contains the software bus message.
+**                         This must point to the first byte of the message header.
+**/
+void CFE_MSG_GenerateChecksum(CFE_SB_MsgPtr_t MsgPtr);
```

Comment:
I'm wondering also, if we are funneling all header operations through API's, methinks the checksum should be automagic? I suppose you could argue that there's a performance hit to calculating every time a field is changed, but I don't know how many times I've written code that doesn't set the checksum.

Reply:
Indeed, when a packet format (container) is defined then a field can be separately designated as a checksum/error control field, so the framework can calculate it and the application software need not be concerned with it or really even know it exists.

Error control fields are mainly to cover data in transit between modules, it isn't worthwhile to (re-)calculate the checksum after every small data change.  App should just fill in _all_ data, and the framework should calculate the checksum/error control field at the time it is "sent", this is when you know the data shouldn't be changing anymore.

Reply:
Agree with everything above.  This routine is envisioned as something SB could call when messages are "sent".  Apps never have to worry about it.

Reply:
> Agree with everything above. This routine is envisioned as something SB could call when messages are "sent". Apps never have to worry about it.

So would I be correct in understanding that we agree that all of the checksum API calls should be removed from the fsw/cfe-core/inc/*.h files?

Although I think there might be value in having utility functions to compare two messages to see if they're identical (whether by computing then checking the checksum or checking the entire message header/body via memcmp.)

Reply:
Incidentally, the proposed change that I had done (see PR #716) was basically this approach - provide a typedef for the error control value, a getter function, a compute function, and a compare function, all based on that type.

The other thing to consider is that the checksum actually changes (depending on algorithm) depending on whether the message needs to be padded for conformance with the CPU alignment requirements.

EDS, of course, handles all of this, if we choose to adopt it - checksums are computed once the message is in its final form for network transmission, using the EDS-specified algorithm.